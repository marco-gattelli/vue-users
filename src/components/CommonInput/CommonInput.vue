<template>
  <input
    type="text"
    class="rounded border px-2 py-1 w-80"
    :class="{
      'bg-gray-100 cursor-not-allowed': disabled,
    }"
    :disabled="disabled"
    :placeholder="placeholder"
    :value="modelValue"
    :data-testid="dataTestId"
    @input="handleInputChange"
  />
</template>

<script setup lang="ts">
interface Props {
  modelValue: string;
  placeholder: string;
  dataTestId?: string;
  disabled?: boolean;
}
const { modelValue, placeholder, disabled, dataTestId } = defineProps<Props>();
const emits = defineEmits<{
  (e: "update:modelValue", modelValue: string): void;
}>();

function handleInputChange($event: Event) {
  emits("update:modelValue", ($event.target as HTMLInputElement).value);
}
</script>
