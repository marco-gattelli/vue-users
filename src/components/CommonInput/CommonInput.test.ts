import { describe, expect, it } from "vitest";
import { mount } from "@vue/test-utils";
import CommonInput from "@/components/CommonInput/CommonInput.vue";

describe("CommonInput", () => {
  it("should render without crash", function () {
    // Arrange.
    // Act.
    const { element } = mount(CommonInput, {
      props: {
        dataTestId: "baz",
        modelValue: "foo",
        placeholder: "bar",
        disabled: false,
      },
    });

    // Assert.
    expect(element).toMatchSnapshot();
  });

  it("should render the input disabled", async function () {
    // Arrange.
    // Act.
    const wrapper = mount(CommonInput, {
      props: {
        dataTestId: "foo",
        modelValue: "foo",
        placeholder: "bar",
        disabled: true,
      },
    });

    // Assert.
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should emit the update event when the input value change", async function () {
    // Arrange.
    const inputDataTestId = "baz";
    const newValue = "foofoo";
    const wrapper = mount(CommonInput, {
      props: {
        dataTestId: inputDataTestId,
        modelValue: "foo",
        placeholder: "bar",
        disabled: false,
      },
    });
    const input = wrapper.find(`[data-testid="${inputDataTestId}"]`);

    // Act.
    await input.setValue(newValue);

    // Assert.
    expect(wrapper.emitted()).toHaveProperty("update:modelValue");
    // @ts-ignore
    expect(wrapper.emitted("update:modelValue")[0][0]).toBe(newValue);
  });
});
