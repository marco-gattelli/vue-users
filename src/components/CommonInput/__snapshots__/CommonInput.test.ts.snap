// Vitest Snapshot v1

exports[`CommonInput > should render the input disabled 1`] = `
<input
  class="rounded border px-2 py-1 w-80 bg-gray-100 cursor-not-allowed"
  data-testid="foo"
  disabled=""
  placeholder="bar"
  type="text"
/>
`;

exports[`CommonInput > should render without crash 1`] = `
<input
  class="rounded border px-2 py-1 w-80"
  data-testid="baz"
  placeholder="bar"
  type="text"
/>
`;
